File: scala/src/test/scala/sodium/MemoryTest1.scala
package sodium

object MemoryTest1 {
  def main(args: Array[String]): Unit = {
    new Thread() {
      override def run(): Unit = {
        try {
          while (true) {
            println(s"memory ${Runtime.getRuntime().totalMemory()}")
            Thread.sleep(5000)
          }
        } catch {
          case e: InterruptedException => e.printStackTrace
        }
      }
    }.start()

    val et = new StreamSink[Int]()
    val t = et.hold(0)
    et.map(x => x / 10)
    val changeTens =
      Stream.filterOptional(
        et.snapshot[Int, Option[Int]](t, (neu: Int, old: Int) => if (neu.equals(old)) None else Some(neu)))
    val oout =
      changeTens.map(tens => t.map(tt => (tens, tt))).hold(t.map(tt => (0, tt)))
    val out = Cell.switchC(oout)
    val l = out.listen(tu => {
      println(s"${tu._1},${tu._2}")
    })
    var i = 0
    while (i < 1000000000) {
      et.send(i)
      i += 1
    }
    l.unlisten()
  }
}

File: java/src/test/java/nz/sodium/MemoryTest1.java
package nz.sodium;

import java.util.Optional;

public class MemoryTest1
{
    public static void main(String[] args)
    {
        new Thread() {
            public void run()
            {
                try {
                    while (true) {
                        System.out.println("memory "+Runtime.getRuntime().totalMemory());
                        Thread.sleep(5000);
                    }
                }
                catch (InterruptedException e) {
                    System.out.println(e.toString());
                }
            }
        }.start();

        StreamSink<Integer> et = new StreamSink<Integer>();
        Cell<Integer> t = et.hold(0);
        Stream<Integer> etens = et.map(x -> x/10);
        Stream<Integer> changeTens = Stream.filterOptional(et.snapshot(t, (neu, old) ->
            neu.equals(old) ? Optional.empty() : Optional.of(neu)));
        Cell<Cell<Tuple2<Integer,Integer>>> oout =
            changeTens.map(tens -> t.map(tt -> new Tuple2<Integer,Integer>(tens, tt))).
            hold(t.map(tt -> new Tuple2<Integer,Integer>(0, tt)));
        Cell<Tuple2<Integer,Integer>> out = Cell.switchC(oout);
        Listener l = out.listen(tu -> {
            //System.out.println(tu.a+","+tu.b);
        });
        int i = 0;
        while (i < 1000000000) {
            et.send(i);
            i++;
        }
        l.unlisten();
    }
}
